README
INSTALLATION
 - Installation instructions can be found in the INSTALL file

TERMINOLOGY

 - Adam
      : new (random) critter

 - Adam Distance (ad)
      : generations with mutations since adam

OPTIONS

  option                                                    [default] [range]      [comment]

  --autoload                                                  0         0-1          autoload critters from ~/.critterding/load
  --benchmark                                                 0         0-1          run the critterding benchmark             
  --body_maxbodyparts                                         30        0-1000000    maximum body parts per critter            
  --body_maxbodypartsatbuildtime                              6         1-1000000    maximum body parts for a new critter      
  --body_maxbodypartsize                                      200       1-1000000    maximum size of a critters body part      
  --body_maxheadsize                                          80        1-1000000    maximum size of a critters head           
  --body_maxmutations                                         3         1-1000000    maximum mutations on a body mutant        
  --body_minbodypartsize                                      20        1-1000000    minimum size of a critters body part      
  --body_minheadsize                                          30        1-1000000    minimum size of a critters head           
  --body_mutationrate                                         10        0-100        percentage of newborns that mutate bodies 
  --body_percentmutateeffectaddbodypart                       1         0-100        chance of adding a body part              
  --body_percentmutateeffectchangecolor                       1         0-100        chance of changing body color             
  --body_percentmutateeffectchangecolor_slightly              5         0-100        chance of changing body color             
  --body_percentmutateeffectchangeconstraintangles            1         0-100        chance of changing a joints position angles
  --body_percentmutateeffectchangeconstraintangles_slightly   5         0-100        chance of changing a joints position angles
  --body_percentmutateeffectchangeconstraintlimits            1         0-100        chance of changing a joints motion limits  
  --body_percentmutateeffectchangeconstraintlimits_slightly   5         0-100        chance of slightly changing a joints motion limits
  --body_percentmutateeffectchangeconstraintposition          1         0-100        chance of changing a joints position              
  --body_percentmutateeffectchangeconstraintposition_slightly 5         0-100        chance of slightly changing a joints position     
  --body_percentmutateeffectremovebodypart                    1         0-100        chance of removing a body part                    
  --body_percentmutateeffectrepositionhead                    5         0-100        chance of repositioning head                      
  --body_percentmutateeffectresizebodypart                    1         0-100        chance of resizing a body part                    
  --body_percentmutateeffectresizebodypart_slightly           5         0-100        chance of slightly resizing a body part           
  --body_percentmutateeffectresizehead                        1         0-100        chance of resizing a head                         
  --body_percentmutateeffectresizehead_slightly               5         0-100        chance of slightly resizing a head                
  --brain_costfiringmotorneuron                               100       0-1000000    cost of firing a motor neuron                     
  --brain_costfiringneuron                                    10        0-1000000    cost of firing a neuron                           
  --brain_costhavingneuron                                    50        0-1000000    cost of having a neuron (1/100000 energy)         
  --brain_costhavingsynapse                                   1         0-1000000    cost of having a synapse                          
  --brain_maxdendridicbranches                                3         1-1000000    maximum number of dendrites per neuron            
  --brain_maxfiringthreshold                                  10        1-1000000    maximum firingthreshold of a neuron               
  --brain_maxmutations                                        10        1-1000000    maximum mutations on a brain mutant               
  --brain_maxneurons                                          1000      1-1000000    maximum neurons per critter                       
  --brain_maxneuronsatbuildtime                               200       1-1000000    maximum neurons for a new critter                 
  --brain_maxplasticitystrengthen                             1000      1-1000000    maximum weight by which plastic synapses strengthen
  --brain_maxplasticityweaken                                 10000     1-1000000    maximum weight by which plastic synapses weaken    
  --brain_maxsynapses                                         100       1-1000000    maximum synapses per neuron                        
  --brain_maxsynapsesatbuildtime                              40        1-1000000    maximum synapses for a new neuron of a new critter 
  --brain_minfiringthreshold                                  2         1-1000000    minimum firingthreshold of a neuron                
  --brain_minneuronsatbuildtime                               50        1-1000000    minimum neurons for a new critter                  
  --brain_minplasticitystrengthen                             100       1-1000000    minimum weight by which plastic synapses strengthen
  --brain_minplasticityweaken                                 1000      1-1000000    minimum weight by which plastic synapses weaken    
  --brain_minsynapses                                         1         1-1000000    minimum synapses per neuron                        
  --brain_minsynapsesatbuildtime                              1         1-1000000    minimum synapses for a new neuron                  
  --brain_mutate_maxdendridicbranches                         0         0-1          mutate this value
  --brain_mutate_maxfiringthreshold                           0         0-1          mutate this value
  --brain_mutate_minfiringthreshold                           0         0-1          mutate this value
  --brain_mutate_mutateeffects                                0         0-1          mutate mutation effects
  --brain_mutate_percentchanceconsistentsynapses              0         0-1          mutate this value
  --brain_mutate_percentchanceinhibitoryneuron                0         0-1          mutate this value
  --brain_mutate_percentchanceinhibitorysynapses              0         0-1          mutate this value
  --brain_mutate_percentchancemotorneuron                     0         0-1          mutate this value
  --brain_mutate_percentchanceplasticneuron                   0         0-1          mutate this value
  --brain_mutate_percentchancesensorysynapse                  0         0-1          mutate this value
  --brain_mutate_plasticityfactors                            0         0-1          mutate min/max plasticity values
  --brain_mutationrate                                        10        0-100        percentage of newborns that mutate brains
  --brain_percentchanceconsistentsynapses                     0         0-100        percent chance a neurons synapses are all inhibitory or excitatory
  --brain_percentchanceinhibitoryneuron                       50        0-100        percent chance a neuron is inhibotory
  --brain_percentchanceinhibitorysynapses                     50        0-100        percent chance a synapse is inhibitory
  --brain_percentchancemotorneuron                            50        0-100        percent chance a neuron is a motor neuron
  --brain_percentchanceplasticneuron                          20        0-100        percent chance a neuron has plastic synapses
  --brain_percentchancesensorysynapse                         20        0-100        percent change a synapse connects to an input
  --brain_percentmutateeffectaddneuron                        1         0-100        chance of adding a neuron
  --brain_percentmutateeffectaddsynapse                       5         0-100        chance of adding a synapse
  --brain_percentmutateeffectaltermutable                     1         0-100        mutate value of a mutatable option
  --brain_percentmutateeffectalterneuron                      2         0-100        chance of altering a neuron
  --brain_percentmutateeffectremoveneuron                     1         0-100        chance of removing a neuron
  --brain_percentmutateeffectremovesynapse                    5         0-100        chance of removing a synapse
  --camerasensitivity                                         20        1-1000       sensitivity of the camera
  --colormode                                                 0         0-1          colors genetically exact critters identically
  --critter_autoexchangeinterval                              0         0-1000000    save critters every n seconds
  --critter_autosaveinterval                                  0         0-1000000    save critters every n seconds
  --critter_enableomnivores                                   1         0-1          enables critters to eat each other
  --critter_insertevery                                       0         0-1000000    inserts a random critter every n frames
  --critter_killhalfat                                        120       2-1000000    kill 50% of critters if population reaches n
  --critter_maxenergy                                         5000      1-1000000    maximum amount of energy a critter has
  --critter_maxlifetime                                       40000     1-1000000    maximum number of frames a critter lives
  --critter_minenergyproc                                     3000      1-1000000    energy a critters needs to procreate
  --critter_procinterval                                      20        1-1000000    minimum frames between procreations
  --critter_raycastvision                                     0         0-1          use raycast vision instead of opengl
  --critter_retinasize                                        8         1-1000       size of a critters eye retina
  --critter_sightrange                                        70        1-1000000    distance a critter can see (10 = 1 worldsize)
  --critter_startenergy                                       3000      1-1000000    energy a new critter (adam) starts with
  --drawscene                                                 1         0-1          draw the scene
  --energy                                                    400       0-1000000    energy in the system by number of food cubes
  --exit_if_empty                                             0         0-1          exit simulation if there are no critters
  --food_maxenergy                                            1500      1-1000000    maximum amount of energy a food unit has
  --food_maxlifetime                                          10000     1-1000000    maximum number of frames a food unit exists
  --food_size                                                 200       1-1000000    size of a food unit
  --fpslimit                                                  30        1-1000       frames per second for the fps limiter
  --fsX                                                       800       1-1000000    fullscreen resolution X
  --fsY                                                       600       1-1000000    fullscreen resolution Y
  --fullscreen                                                0         0-1          enable fullscreen mode
  --headless                                                  0         0-1          do not open gl context
  --killhalf_decrenergypct                                    1         0-100        decrease energy by n percent when killhalfat triggers
  --killhalf_decrmaxlifetimepct                               0         0-100        decrease critter_maxlifetime by n when killhalfat triggers
  --killhalf_incrworldsizeX                                   0         0-100        increase worldsizeX by n when killhalfat triggers
  --killhalf_incrworldsizeY                                   0         0-100        increase worldsizeY by n when killhalfat triggers
  --mincritters                                               10        0-1000       minimum number of critters
  --race                                                      0         0-1          enable race simulation
  --retinasperrow                                             20        1-1000       number of vision retinas to stack per row onscreen
  --roundworld                                                0         0-1          enable round planet
  --startseed                                                 0         0-4000000000 enable fullscreen mode
  --testworld                                                 0         0-1          a world for test purposes
  --threads                                                   1         1-16         threads to use
  --worldsizeX                                                23        1-5000       size of the world along axis X
  --worldsizeY                                                13        1-5000       size of the world along axis Y
  --worldwalls                                                1         0-1          enable walls around the world

 To save the default settings to a profile, press "s" in the simulation.
 It will be saved to ./default and can be loaded by using "--profile ./default"

 Use F1 in the simulation for more information about keys.


OTHER THINGS

 - Make sure the retina boxes in the lower left of the window never leave the
   boundaries of the window.  Offscreen rendering is not supported by many
   video card drivers.  For some drivers vision will not work if you minimize
   the window or change desktops.


Bob Winckelmans <bob.winckelmans@telenet.be>
IRC: #critterding@irc.freenode.org

Changelog
1.0-beta12.1
------------
catch segfault and warn when using headless without critter_raycastvision
some fixes and cleanups

1.0-beta12
----------
critter selection and actions (kill, duplicate)
new panels: neural net brainviewer, hud, species list
touchingcritter brain input fix
profile saves to ~/critterding/save/(profile)/(profile).pro
toggle rendering of gui and scene (h & r)
font switch to DejaVuSans
an icon thanks to jrabbit
new options:
 --roundworld: a round planet
 --benchmark: times a scene of 10000 frames
 --headless: console mode
 --startseed: seed for the random number generator
 --killhalf_incrworldsizeX/Y: option to increase worldsize when killhalf triggers
 --killhalf_decrmaxlifetimepct: option to decrease critter maxlifetime when killhalf triggers
 --critter_raycastvision: raycast vision for critters
 --threads, number of threads, making openmp (comes with gcc4.2+) a requirement
build system: 
 if available, build against system ftgl (pass --disable-system-ftgl to use internal)
 make install
many fixes, cleanups & changes

1.0-beta11
---------
improved keyboard / mouse input handling
improved gui with panels for realtime setting manipulation,
critters vs food graph,
exit panel (esc no longer exits)
a few more body mutations
fix: critter autosaving timer
fix: crashes on exit
many more fixes and cleanups

1.0-beta10
---------
changed the physics backend to Bullet Physics
changed input and window backend to SDL
a racing mode
mouse support for looking and interactivity
profile saving (patch from Eric Burton)
fullscreen support
many improvements, fixes and cleanups overall

1.0-beta9
---------
dedicated skincolors for herbivores(Red/Blue) / carnivores(Green/Blue)
bullets no longer shoot through walls
world is no longer fixed to a square
fps limiter
introduced text rendering with fglx
improved synaptic plasticity
slightly improved wall system
--autoload option
cleaned up critter files
color inputneurons get less discrete input
many fixes and cleanups

1.0-beta8
---------
new: custom settings profiles, critter autosaving, exit-if-empty
corpses and carrying are disabled at default
various fixes and cleanups

1.0-beta7
---------
critters are now classified as herbivore or carnivore
corpse is poison to a herbivore/green food is poison to a carnivore
fixed a alpha color bug in critter vision (vidia cards)
engine improvements: backface & frustrum culling + display lists
more efficient copying of brains at procreation time
simplified (more usable) camera with adjustable sensitivity
critters can now pick up and drop food

1.0-beta6
---------
age input neurons
started testing a corpse entity
discovered and fixed retina to input mapping bug
visibility enhancement by allowing only bright colors
bullet hit: half of lifetime lost
camera autopositioning according to world size
--small, --medium, --big and --huge options

1.0-beta5
---------
Options have boundary checks now
improved command line option handling
f1 pauses and prints settings / keys
increased evolution rate by tuned defaults settings
fixed README mistakes

1.0-beta4
---------
adaptable colorneurons, default is now 3 instead of 2
adaptable retinasize, default is nog 7 instead of 10
all (38) initial brain settings have been made adaptable
provided an option to flip newborn critters 180 degrees
simple --help implemented
many minor fixes and cleanups

1.0-beta3
---------
fixes minor memleaks & speedups
brain.clearInputs() implementation                             
critter & food size have been made adaptable
make sure we only load files with extension cr
Retinas Per Row is now adaptable
default critter sight distance reduced from 50 to 40: performance boost
follow camera handles critter sight distance
cleaned up settings reporting
updated README
Start of a Changelog


ocr/ocr.cpp
//	mplayer tv:// -tv driver=v4l:width=160:height=120:device=/dev/video1:outfmt=rgb32:noaudio -fps 3 -vf scale=80:60,format=bgr15 -vo tga

#include <iostream>
#include <cstdlib>
#include <iomanip>

#include "../utils/dir.h"
#include "../utils/file.h"
#include <tgastore.h>
#include <brainstore.h>

using namespace std;

int main(int argc, char *argv[])
{
	// load stores
		tgastore *picstore = new tgastore;
		brainstore *bstore = new brainstore;

	// get a file and directory handle
		Dir			dirH;
		File			fileH;

	// determine home / program directory
		string homedir = getenv("HOME");
		if ( homedir.empty() ) {
			cout << "environment variable HOME not defined/detected" << endl;
			exit(0);
		}
		string progdir = homedir;	progdir.append("/.m5");
		if ( !dirH.exists(progdir) )	dirH.make(progdir);

	// load pictures
		cerr << "loading pictures" << endl;

		vector<string> tgafiles;

		dirH.listContentsFull("ocr/pics/empty_desk/", tgafiles);
		unsigned int empty_desk = tgafiles.size();
		for ( unsigned int i = 0; i < tgafiles.size(); i++ )
			picstore->load(tgafiles[i].c_str());

		tgafiles.clear();
		dirH.listContentsFull("ocr/pics/me_at_desk/", tgafiles);
		unsigned int me_at_desk = tgafiles.size();
		for ( unsigned int i = 0; i < tgafiles.size(); i++ )
			picstore->load(tgafiles[i].c_str());

		tgafiles.clear();
		dirH.listContentsFull("ocr/pics/open_door/", tgafiles);
		unsigned int open_door = tgafiles.size();
		for ( unsigned int i = 0; i < tgafiles.size(); i++ )
			picstore->load(tgafiles[i].c_str());

		tgafiles.clear();
		dirH.listContentsFull("ocr/pics/night/", tgafiles);
		unsigned int night = tgafiles.size();
		for ( unsigned int i = 0; i < tgafiles.size(); i++ )
			picstore->load(tgafiles[i].c_str());

	// upload questions & expected answers
		for ( unsigned int p = 0; p < picstore->pictures.size(); p++ )
		{
			unsigned int expect	= 0;

			if ( p < empty_desk )							expect = 0;
			else if ( p < empty_desk+me_at_desk )					expect = 1;
			else if ( p < empty_desk+me_at_desk+open_door )				expect = 2;
			else if ( p < empty_desk+me_at_desk+open_door+night )			expect = 3;

			cerr << "QUESTION PICTURE: " << p << " expects: " << expect << endl;
			bstore->setQuery(&picstore->pictures[p], expect);
		}


//exit(0);

	// create initial (loaded) brains

		cerr << "loading initial brains" << endl;
		unsigned int nbrains = 58;
		unsigned int selectbrains = 29;
		bstore->selectBrains = selectbrains;

		string loaddir = progdir;	loaddir.append("/load");

// 		vector<string> files;
// 		dirH.listContentsFull(loaddir, files);
// 		for ( unsigned int i = 0; i < selectbrains; i++ )
// 		{
// 			cout << "loading " << files[0] << endl;
// 			string content;
// 			fileH.open( files[0], content );
//  
// 			bstore->load( &content );
// 		}

	// create initial fresh brains
		unsigned int inputs = picstore->pictures[0].size() * 10;
		unsigned int outputs = 4;
		for (unsigned int i=0; i < selectbrains; i++)
			bstore->load( inputs, outputs );

	// saving parameters
		unsigned int savedCritters = 0;
		unsigned int saveCounter = 0;
		unsigned int saveEvery = 100;

		// determine and create directories
		string savedir = progdir;	savedir.append("/save");
		stringstream buf;
			buf << savedir << "/" << time(0);
		string subsavedir = buf.str();

		if ( !dirH.exists(savedir) )	dirH.make(savedir);
		if ( !dirH.exists(subsavedir) )	dirH.make(subsavedir);

	unsigned int generations = 0;
	while (1)
	{
		// process ALL pictures by all critters
	
			bstore->process( );
	
		// initialize indices
			vector<int> indices ( nbrains, 0 );
			for ( unsigned int i = 0; i < nbrains; i++ )
				indices[i] = i;
	
		// sort results
			for ( int i = nbrains; i>0; i--  )
				for ( int j = 0; j < i-1; j++  )
					if ( bstore->brainusers[indices[j]]->weightedresult > bstore->brainusers[indices[j+1]]->weightedresult )
					{
						unsigned keepI	= indices[j];
						indices[j]	= indices[j+1];
						indices[j+1]	= keepI;
					}
	
		// print out SORTED results
			generations++;
			system("clear");
			for (unsigned int i=0; i < nbrains; i++)
				cerr << "gen:" << setw(6) << generations << " brain:" << setw(4) << indices[i] << " : " << "right: " << setw(4) << bstore->brainusers[indices[i]]->results << "/" << picstore->pictures.size() << ", wrong: " << setw(3) << bstore->brainusers[indices[i]]->misses << " (N: " << setw(4) << bstore->brainusers[indices[i]]->brain.totalNeurons << " C: " << setw(5) << bstore->brainusers[indices[i]]->brain.totalSynapses << " F: " << setw(5) << bstore->brainusers[indices[i]]->brain.neuronsFired << ") " << setprecision(10) << setw(10) << bstore->brainusers[indices[i]]->weightedresult << endl;
			//cerr << endl;

		// save best Critter
			if ( ++saveCounter == saveEvery )
			{
				saveCounter = 0;
	
				string* bstring = bstore->brainusers[indices[nbrains-1]]->brain.getArch();
	
				cerr << "saving brain " << indices[nbrains-1] << endl;
	
				// determine filename
				stringstream filename;
				filename << subsavedir << "/brain-" << setfill('0') << setw(6) << savedCritters;
				string sfilename = filename.str();
	
				// save brain
				fileH.save(sfilename, bstring);
	
				savedCritters++;
			}
	
		// remove worst ones
			for ( unsigned int i=0; i < selectbrains; i++ )
			{
				//cerr << "removing " << indices[0] << endl;
				delete bstore->brainusers[indices[0]];
				bstore->brainusers.erase(bstore->brainusers.begin()+indices[0]);
	
				// adapt indices, higher ones drop one
				for ( unsigned int k=1; k < indices.size(); k++ )
					if ( indices[k] > indices[0] ) indices[k]--;
	
				indices.erase(indices.begin());
			}

	}


	// destruct
	delete picstore;
	delete bstore;
}
